### ISSUE TITLE
type class for controlling dropout

### ISSUE THREAD

**Issue by tscholak** [hasktorch/hasktorch#213]
In pytorch, one can call `model.train()` to turn all dropout layers on for training, and one can call `model.eval()` to turn all dropout layers off for evaluation.

For the transformer implementation, I opted for exposing a `Bool` in every forward function that regulates that.

However, maybe it is better to put that `Bool` value into the `NN` data type and make a type class that, similarly to `Parameterized` offers a recursively defined function that can be used to flip the switch between training and evaluation for all dropout layers in the model hierarchy.

thoughts?

**Comment by o1lo01ol1o** [comment]
FWIW, I would not couple `Dropout` to concepts of training or evaluation [since it's reasonable that one might want dropout in circumstances that are not part of training.](https://arxiv.org/pdf/1506.02142.pdf)

However, all the apis I'm aware of leave something to be desired on this front and I'm afraid I don't have any useful ideas.

**Comment by tscholak** [comment]
@o1lo01ol1o oh, for sure. what I wrote above would just be a way to flip the switch between dropout turned on and dropout turned off. when it's used and how is up to the user of the lib.

**Comment by apaszke** [comment]
We could also try being fancy and go e.g. down the algebraic effects path, and have training/dropout be different interpreters for the sampling statements 🤔 But we would have to be careful about the way in which we expose this functionality to avoid making it too complex.
